certs/certs.go

```diff
@@ -0,0 +1,293 @@
+// `MakePowerTableDiff`) and justification (from GPBFT).
+//
+// Note, however, that this function does not attempt to validate the resulting finality
+// certificate (beyond verifying that it is a justification for the correct round). You can do so by
+// immediately calling `ValidateFinalityCertificates` on the result.
+func NewFinalityCertificate(powerDelta []PowerTableDelta, justification *gpbft.Justification) (*FinalityCertificate, error) {
+	if justification.Vote.Step != gpbft.DECIDE_PHASE {
+		return nil, xerrors.Errorf("can only create a finality certificate from a decide vote, got phase %s", justification.Vote.Step)
+	}
+
+	if justification.Vote.Round != 0 {
+		return nil, xerrors.Errorf("expected decide round to be 0, got round %d", justification.Vote.Round)
+	}
+
+	if justification.Vote.Value.IsZero() {
+		return nil, xerrors.Errorf("got a decision for bottom for instance %d", justification.Vote.Instance)
+	}
+
+	return &FinalityCertificate{
+		GPBFTInstance:    justification.Vote.Instance,
+		SupplementalData: justification.Vote.SupplementalData,
+		ECChain:          justification.Vote.Value,
+		Signers:          justification.Signers,
+		Signature:        justification.Signature,
+		PowerTableDelta:  powerDelta,
+	}, nil
+}
+
+// ValidateFinalityCertificates validates zero or more finality certificates, returning the next
+// instance after the last valid finality certificates, any newly finalized tipsets, and the next
+// power table to use as-of the last valid finality certificate. If passed a non-nil `base` tipset,
+// validate that the finalized chain starts with that tipset (accept any finalized chain otherwise).
+//
+// Returns an error if it encounters any invalid finality certificates, along with the last valid
+// instance, finalized chain epochs, etc. E.g., if provided a partially valid chain of finality
+// certificates, this function will return a (possibly empty) prefix of the EC chain correctly
+// finalized, the instance of the first invalid finality certificate, and the power table that
+// should be used to validate that finality certificate, along with the error encountered.
+func ValidateFinalityCertificates(verifier gpbft.Verifier, network gpbft.NetworkName, prevPowerTable gpbft.PowerEntries, nextInstance uint64, base *gpbft.TipSet,
+	certs ...FinalityCertificate) (_nextInstance uint64, chain gpbft.ECChain, newPowerTable gpbft.PowerEntries, err error) {
+	for _, cert := range certs {
+		if cert.GPBFTInstance != nextInstance {
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("expected instance %d, found instance %d", nextInstance, cert.GPBFTInstance)
+		}
+		// Some basic sanity checks.
+		if err := cert.ECChain.Validate(); err != nil {
+			return nextInstance, chain, prevPowerTable, xerrors.Errorf("invalid finality certificate at instance %d: %w", cert.GPBFTInstance, err)
+		}
+
+		// We can't have a finality certificate for "bottom"
+		if cert.ECChain.IsZero() {
```

Comment:
Is that right? I'm not certain. I'm pretty sure the GPBFT can terminate with a bottom result.

Reply:
Or will the justification be for the Base?

Reply:
From what I can tell, we cannot "decide" zero: https://github.com/filecoin-project/go-f3/blob/4f6464d9c33fae650918347bcdb06b5911922e5b/gpbft/gpbft.go#L430-L432

We can commit to zero to start another round, but we'll keep going until we decide on _at least_ the base.